In [2]:
import pandas as pd
df = pd.read_csv('housepricedata.csv')#Reading the house price data

In [3]:
df

LotArea  OverallQual  OverallCond  TotalBsmtSF  FullBath  HalfBath  \
0        8450            7            5          856         2         1   
1        9600            6            8         1262         2         0   
2       11250            7            5          920         2         1   
3        9550            7            5          756         1         0   
4       14260            8            5         1145         2         1   
5       14115            5            5          796         1         1   
6       10084            8            5         1686         2         0   
7       10382            7            6         1107         2         1   
8        6120            7            5          952         2         0   
9        7420            5            6          991         1         0   
10      11200            5            5         1040         1         0   
11      11924            9            5         1175         3         0   
12      12968            5            6          912         1         0   
13      10652            7            5         1494         2         0   
14      10920            6            5         1253         1         1   
15       6120            7            8          832         1         0   
16      11241            6            7         1004         1         0   
17      10791            4            5            0         2         0   
18      13695            5            5         1114         1         1   
19       7560            5            6         1029         1         0   
20      14215            8            5         1158         3         1   
21       7449            7            7          637         1         0   
22       9742            8            5         1777         2         0   
23       4224            5            7         1040         1         0   
24       8246            5            8         1060         1         0   
25      14230            8            5         1566         2         0   
26       7200            5            7          900         1         0   
27      11478            8            5         1704         2         0   
28      16321            5            6         1484         1         0   
29       6324            4            6          520         1         0   
...       ...          ...          ...          ...       ...       ...   
1430    21930            5            5          732         2         1   
1431     4928            6            6          958         2         0   
1432    10800            4            6          656         2         0   
1433    10261            6            5          936         2         1   
1434    17400            5            5         1126         2         0   
1435     8400            6            9         1319         1         1   
1436     9000            4            6          864         1         0   
1437    12444            8            5         1932         2         0   
1438     7407            6            7          912         1         0   
1439    11584            7            6          539         2         1   
1440    11526            6            7          588         2         0   
1441     4426            6            5          848         1         0   
1442    11003           10            5         1017         2         1   
1443     8854            6            6          952         1         0   
1444     8500            7            5         1422         2         0   
1445     8400            6            5          814         1         0   
1446    26142            5            7         1188         1         0   
1447    10000            8            5         1220         2         1   
1448    11767            4            7          560         1         1   
1449     1533            5            7          630         1         0   
1450     9000            5            5          896  

### In last column the colum AboveMedianPrice means that if the house is above the median price or not. Here 1 means yes and 0 means no.

In [6]:
#converting in the arrays so thatour machine could process
dataset = df.values
dataset

array([[ 8450,     7,     5, ...,     0,   548,     1],
       [ 9600,     6,     8, ...,     1,   460,     1],
       [11250,     7,     5, ...,     1,   608,     1],
       ...,
       [ 9042,     7,     9, ...,     2,   252,     1],
       [ 9717,     5,     6, ...,     0,   240,     0],
       [ 9937,     5,     6, ...,     0,   276,     0]])

As we can see the our values are stored in the array now.

Now we will split our dataset into input features (X) and the feature we wish to predict (Y).


In [7]:
X = dataset[:,0:10]

Above code before comma means that we are not splitting the rows and after the comma means we are extracting the first 10 columns therefore using the '0:10' which will take the value from 0 to 9 and will put it in the variable X

In [9]:
#assigning the last column of array to Y
Y = dataset[:,10]

In [10]:
#Ensuring that scaling of our input features are similar with the help of sckit-learn
from sklearn import preprocessing

#using the function 'min-max scaler' which scales the dataset so that
#all the features lie between 0 and 1
min_max_scaler = preprocessing.MinMaxScaler()
X_scale = min_max_scaler.fit_transform(X)
X_scale

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:475: DataConversionWarning: Data with input dtype int64 was converted to float64 by MinMaxScaler.
  warnings.warn(msg, DataConversionWarning)


array([[0.0334198 , 0.66666667, 0.5       , ..., 0.5       , 0.        ,
        0.3864598 ],
       [0.03879502, 0.55555556, 0.875     , ..., 0.33333333, 0.33333333,
        0.32440056],
       [0.04650728, 0.66666667, 0.5       , ..., 0.33333333, 0.33333333,
        0.42877292],
       ...,
       [0.03618687, 0.66666667, 1.        , ..., 0.58333333, 0.66666667,
        0.17771509],
       [0.03934189, 0.44444444, 0.625     , ..., 0.25      , 0.        ,
        0.16925247],
       [0.04037019, 0.44444444, 0.625     , ..., 0.33333333, 0.        ,
        0.19464034]])

In [13]:
#We will now split our data into the training set and test set with the help of the sklearn
from sklearn.model_selection import train_test_split
X_train, X_val_and_test, Y_train, Y_val_and_test = train_test_split(X_scale, Y, test_size=0.3)
#This tell sklearn that your_val_and_test size will be 30% of the overall dataset.

#using the same function to do the split again on val_and_test
X_val, X_test, Y_val, Y_test = train_test_split(X_val_and_test, Y_val_and_test, test_size=0.5)

print(X_train.shape, X_val.shape, X_test.shape, Y_train.shape, Y_val.shape, Y_test.shape)

(1022, 10) (219, 10) (219, 10) (1022,) (219,) (219,)


### As we can see, We have 1022 data points while the validation and test set has 219 data points each.

#### Till now we have done following things step-wise
1. Read in the CSV file
2. Split our dataset into the input features and the label
3. Scale the data so that the input features have similar orders of the magnitude.
4. Splitting our dataset into the training set, the validation set and the test set.

# Building And Training The Neural Network

In [14]:
from keras.models import Sequential
from keras.layers import Dense

/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


Since we want to have have follwing kind of layers in the Neural Network
1. Hidden Layer 1:32 neurons, ReLU activation
2. Hidden Layer 2:32 neurons, ReLU activation
3. Output Layer 1 neuron, Sigmoid activation

We will run the follwoing code to fulfill our requirements

In [17]:
#This says that we will store our model in the variable 'model' and we will
#describe it sequentially(layer by layer)
model = Sequential([
    Dense(32, activation='relu', input_shape=(10,)),
    Dense(32, activation='relu'),
    Dense(1, activation='sigmoid'),
])

### Filling in the best numbers


In [18]:
model.compile(optimizer='sgd',
             loss='binary_crossentropy',
             metrics=['accuracy'])
#sgd means stochastic gradient descent
#loss function for outputs that thakes the values 1 or 0 is called binary cross entropy
#metrics=accuracy means we want to track accuracy on top of the loss function

In [19]:
#By the function fit we are fitting the parameters to the data
hist = model.fit(X_train, Y_train,
                batch_size=32, epochs=100,
                validation_data = (X_val, Y_val))

Train on 1022 samples, validate on 219 samples
Epoch 1/100
1022/1022 [==============================] - 0s 449us/step - loss: 0.7020 - acc: 0.4912 - val_loss: 0.6971 - val_acc: 0.5023
Epoch 2/100
1022/1022 [==============================] - 0s 48us/step - loss: 0.6948 - acc: 0.5068 - val_loss: 0.6903 - val_acc: 0.5662
Epoch 3/100
1022/1022 [==============================] - 0s 47us/step - loss: 0.6888 - acc: 0.5705 - val_loss: 0.6850 - val_acc: 0.6804
Epoch 4/100
1022/1022 [==============================] - 0s 46us/step - loss: 0.6838 - acc: 0.6967 - val_loss: 0.6804 - val_acc: 0.7717
Epoch 5/100
1022/1022 [==============================] - 0s 47us/step - loss: 0.6794 - acc: 0.7368 - val_loss: 0.6762 - val_acc: 0.7717
Epoch 6/100
1022/1022 [==============================] - 0s 47us/step - loss: 0.6751 - acc: 0.7417 - val_loss: 0.6721 - val_acc: 0.7717
Epoch 7/100
1022/1022 [==============================] - 0s 47us/step - loss: 0.6708 - acc: 0.7583 - val_loss: 0.6680 - val_acc: 0.7626


1022/1022 [==============================] - 0s 49us/step - loss: 0.3274 - acc: 0.8748 - val_loss: 0.3366 - val_acc: 0.8813
Epoch 61/100
1022/1022 [==============================] - 0s 53us/step - loss: 0.3252 - acc: 0.8777 - val_loss: 0.3342 - val_acc: 0.8858
Epoch 62/100
1022/1022 [==============================] - 0s 48us/step - loss: 0.3230 - acc: 0.8757 - val_loss: 0.3319 - val_acc: 0.8858
Epoch 63/100
1022/1022 [==============================] - 0s 47us/step - loss: 0.3208 - acc: 0.8728 - val_loss: 0.3296 - val_acc: 0.8767
Epoch 64/100
1022/1022 [==============================] - ETA: 0s - loss: 0.3212 - acc: 0.906 - 0s 45us/step - loss: 0.3184 - acc: 0.8728 - val_loss: 0.3278 - val_acc: 0.8813
Epoch 65/100
1022/1022 [==============================] - 0s 48us/step - loss: 0.3172 - acc: 0.8728 - val_loss: 0.3260 - val_acc: 0.8950
Epoch 66/100
1022/1022 [==============================] - 0s 45us/step - loss: 0.3152 - acc: 0.8757 - val_loss: 0.3237 - val_acc: 0.8813
Epoch 67/100
102

In [20]:
#Evaluating on the test set
model.evaluate(X_test, Y_test)[1]#

219/219 [==============================] - 0s 43us/step


0.8904109597206116